segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -0,0 +1,483 @@
+                applyTruncate(txn, systemSegment, truncateAt, firstChunkStartsAt);
+            }
+        }
+    }
+
+    /**
+     * Apply chunk addition.
+     */
+    private void applyChunkAddition(MetadataTransaction txn, HashMap<String, Long> chunkStartOffsets, String segmentName, String oldChunkName, String newChunkName, long offset) throws StorageMetadataException {
+        Preconditions.checkState(null != oldChunkName);
+        Preconditions.checkState(null != newChunkName && !newChunkName.isEmpty());
+
+        SegmentMetadata segmentMetadata = (SegmentMetadata) txn.get(segmentName);
+        segmentMetadata.checkInvariants();
+
+        // set length.
+        segmentMetadata.setLength(offset);
+
+        val newChunkMetadata = ChunkMetadata.builder()
+                .name(newChunkName)
+                .build();
+        txn.create(newChunkMetadata);
+        txn.markPinned(newChunkMetadata);
+
+        chunkStartOffsets.put(newChunkName, offset);
+        // Set first and last pointers.
+        if (!oldChunkName.isEmpty()) {
+            ChunkMetadata oldChunk = (ChunkMetadata) txn.get(oldChunkName);
+            Preconditions.checkState(null != oldChunk);
+
+            // In case the old segment store was still writing some zombie chunks when ownership changed
+            // then new offset may invalidate tail part of chunk list.
+            // Note that chunk with oldChunkName is still valid, it is the chunks after this that become invalid.
+            String toDelete = oldChunk.getNextChunk();
+            while (toDelete != null) {
+                ChunkMetadata chunkToDelete = (ChunkMetadata) txn.get(toDelete);
+                txn.delete(toDelete);
+                toDelete = chunkToDelete.getNextChunk();
+            }
+
+            // Set next chunk
+            oldChunk.setNextChunk(newChunkName);
+
+            // Set length
+            long oldLength = chunkStartOffsets.get(oldChunkName);
+            oldChunk.setLength(Math.toIntExact(offset - oldLength));
+
+            txn.update(oldChunk);
+        } else {
+            segmentMetadata.setFirstChunk(newChunkName);
+        }
+        segmentMetadata.setLastChunk(newChunkName);
+        segmentMetadata.setLastChunkStartOffset(offset);
+        segmentMetadata.checkInvariants();
+        // Save the segment metadata.
+        txn.update(segmentMetadata);
+    }
+
+    private String getSystemJournalFileName() {
+        return getSystemJournalFileName(containerId, epoch, currentFileIndex);
+    }
+
+    private String getSystemJournalFileName(int containerId, long epoch, long currentFileIndex) {
+        return NameUtils.getSystemJournalFileName(containerId, epoch, currentFileIndex);
+    }
+
+    private ChunkHandle getChunkHandleForSystemJournal() throws ChunkStorageException {
+        ChunkHandle h;
+        val systemLogName = getSystemJournalFileName();
+        if (!chunkStorage.exists(systemLogName)) {
```

Comment:
Be optimistic. Do `openWrite` and only if it fails with the appropriate exception invoke `create`.

Reply:
At the current moment I need debug friendly easy code.
Once I am done with all changes, I do plan to implement a few more optimization at the very end as part of the last PR in this group.

Reply:
Note that this code is called only during container startup.